modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/near/IgniteTxExceptionNodeFailTest.java

```diff
@@ -0,0 +1,193 @@
+        testNodeFail(PRIMARY_SYNC);
+    }
+
+    /**
+     * Test with {@link CacheWriteSynchronizationMode#FULL_SYNC}
+     * @throws Exception
+     */
+    @Test
+    public void testNodeFailWithFullSync() throws Exception {
+        testNodeFail(FULL_SYNC);
+    }
+
+    /**
+     * <ul>
+     * <li>Start 2 nodes with transactional cache without backups
+     * <li>Start transaction:
+     *  <ul>
+     *  <li>put a key to a partition on transaction coordinator
+     *  <li>put a key to a partition on other node
+     *  <li>try to commit the transaction
+     *  </ul>
+     * <li>Stop other node when it try to send GridNearTxFinishResponse
+     * <li>Check that {@link Transaction#commit()} throw {@link TransactionHeuristicException}
+     * </ul>
+     *
+     * @param testSyncMode
+     * @throws Exception
+     */
+    private void testNodeFail(CacheWriteSynchronizationMode testSyncMode) throws Exception {
+        syncMode = testSyncMode;
+
+        startGrids(2);
+
+        grid(0).cluster().active(true);
+
+        IgniteEx grid0 = grid(0);
+        IgniteEx grid1 = grid(1);
+
+        int key0 = 0;
+        int key1 = 0;
+
+        Affinity<Object> aff = grid1.affinity("cache");
+
+        for (int i = 1; i < 1000; i++) {
+            if (grid0.equals(grid(aff.mapKeyToNode(i)))) {
+                key0 = i;
+
+                break;
+            }
+        }
+
+        for (int i = key0; i < 1000; i++) {
+            if (grid1.equals(grid(aff.mapKeyToNode(i))) && !aff.mapKeyToNode(key1).equals(aff.mapKeyToNode(i))) {
+                key1 = i;
+
+                break;
+            }
+        }
+
+        assert !aff.mapKeyToNode(key0).equals(aff.mapKeyToNode(key1));
+
+        try (Transaction tx = grid1.transactions().txStart()) {
+            grid1.cache("cache").put(key0, 100);
+            grid1.cache("cache").put(key1, 200);
+
+            spi(grid0).blockMessages((node, msg) -> {
+                    if (msg instanceof GridNearTxFinishResponse) {
+                        new Thread(
+                            new Runnable() {
+                                @Override public void run() {
+                                    log().info("Stopping node: [" + grid0.name() + "]");
+
+                                    IgnitionEx.stop(grid0.name(), true, null, true);
+                                }
+                            },
+                            "node-stopper"
+                        ).start();
+
+                        return true;
+                    }
+
+                    return false;
+                }
+            );
+
+            try {
+                tx.commit();
+            }
+            catch (Throwable e) {
+                String msg = e.getMessage();
+
+                Assert.isTrue(e.getCause() instanceof CacheInvalidStateException);
+
+                Pattern msgPattern = Pattern.compile("Failed to commit a transaction \\(all partition owners have left the grid, " +
```

Comment:
If it's possible, extract message in TX code as a public static final constant use reuse it in this test. If someone has to change the message, this test will fail.

Reply:
fixed